### PR TITLE
Arman: Paint

### PR DESCRIPTION
**Pull request by hsp-arman**

### CODE REVIEW

File: src/main/java/am/hsp/apps/paint/exception/InvalidBrushException.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package main.java.am.hsp.apps.paint.exception;
+
+/**
+ * Invalid Brush Exception
+ * @author arman
+ *
+ */
+public class InvalidBrushException extends RuntimeException{
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/paint/exception/InvalidBrushException.java]: Please generate serial version id!

File: src/main/java/am/hsp/apps/paint/toolBox/ToolBox.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+import main.java.am.hsp.apps.paint.toolBox.brush.Brush;
+import main.java.am.hsp.apps.paint.toolBox.brush.BrushExamples;
+import main.java.am.hsp.apps.paint.toolBox.color.Color;
+import main.java.am.hsp.apps.paint.toolBox.color.ColorExamples;
+import main.java.am.hsp.apps.paint.toolBox.shape.Shape;
+import main.java.am.hsp.apps.paint.toolBox.shape.ShapeExamples;
+
+/**
+ * Class ToolBox
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/paint/toolBox/ToolBox.java]: Comments... please fix everywhere

File: src/main/java/am/hsp/apps/paint/toolBox/brush/Brush.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
+package main.java.am.hsp.apps.paint.toolBox.brush;
+
+/**
+ * Class Brush
+ * @author arman
+ *
+ */
+public class Brush {
+    
+    private int brushId;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/paint/toolBox/brush/Brush.java]: remember clean code principles. There is no need to append a class name before each variable name...

File: src/main/java/am/hsp/apps/paint/toolBox/brush/BrushExamples.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import main.java.am.hsp.apps.paint.exception.InvalidBrushException;
+
+/**
+ * Class BrushExamples
+ * @author arman
+ *
+ */
+public class BrushExamples {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/paint/toolBox/brush/BrushExamples.java]: Bad class name! It is better to use BrushPanel for example...

File: src/main/java/am/hsp/apps/paint/toolBox/brush/BrushExamples.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    private static final int MAX_ID = 3;
+
+    private static final int MIN_SIZE = 1;
+    private static final int MID_SIZE = 2;
+    private static final int MAX_SIZE = 3;
+
+    /**
+     * Create list for brushes
+     */
+    public List<Brush> brushes;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/paint/toolBox/brush/BrushExamples.java]: private

File: src/main/java/am/hsp/apps/paint/toolBox/brush/BrushExamples.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+    private static final int MIN_ID = 1;
+    private static final int MID_ID = 2;
+    private static final int MAX_ID = 3;
+
+    private static final int MIN_SIZE = 1;
+    private static final int MID_SIZE = 2;
+    private static final int MAX_SIZE = 3;
+
+    /**
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/paint/toolBox/brush/BrushExamples.java]: No comment needed

File: src/main/java/am/hsp/apps/paint/toolBox/brush/BrushExamples.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    /**
+     * Is brush unavailable in example ?
+     * 
+     * @param brushId
+     * @return
+     */
+    private boolean isBrushUnavailable(int brushId) {
+        for (Brush brush : brushes) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/paint/toolBox/brush/BrushExamples.java]: DRY violation

File: src/main/java/am/hsp/apps/paint/toolBox/shape/Diamond.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    public static final int DIAMOND_ID = 3;
+
+    /**
+     * Method to print diamond
+     * 
+     * @param size
+     */
+    @Override
+    public void draw(Color color, Brush brush) {
+        int size = SIZE_OF_SHAPE;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/paint/toolBox/shape/Diamond.java]: you have this line in each shape class - so why you didn't create a size instance variable in the base class and then use it in sub classes???

File: src/main/java/am/hsp/apps/paint/toolBox/shape/Diamond.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    protected void printBlank(int size) {
+        for (int l = 0; l < size + 1; l++) {
+            System.out.print(" ");
+        }
+    }
+
+    /**
+     * Constructor
+     */
+    public Diamond() {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/paint/toolBox/shape/Diamond.java]: Constructor should be the FIRST methods after instance variables declaration

File: src/main/java/am/hsp/apps/paint/toolBox/shape/Shape.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    public abstract void draw(Color color, Brush brush);
+
+    /**
+     * Create string to brush
+     * 
+     * @param brushSize
+     * @param brush
+     * @return
+     */
+    public StringBuilder brushString(Color color, Brush brush) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/paint/toolBox/shape/Shape.java]: method names MUST BE verbs. For example getFinalSymbol

File: src/main/java/am/hsp/apps/paint/toolBox/shape/Shape.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+        StringBuilder resultString = new StringBuilder(" ");
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/paint/toolBox/shape/Shape.java]: Bad variable name. Actually this is the final symbol which you will use to paint the shapes

File: src/main/java/am/hsp/apps/paint/toolBox/shape/Shape.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * Print a row from symbols in given size
+     * 
+     * @param color
+     * @param brush
+     */
+    public void printSymbols(Color color, Brush brush, int size) {
+
+        for (int j = 0; j < size; j++) {
+            System.out.print(brushString(color, brush));
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/paint/toolBox/shape/Shape.java]: Very bad. So you are calling brushString method in EACH iteration - which always constructs and returns THE SAME final symbol?

File: src/main/java/am/hsp/apps/paint/toolBox/shape/Shape.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+        return resultString;
+    }
+
+    /**
+     * Print a row from symbols in given size
+     * 
+     * @param color
+     * @param brush
+     */
+    public void printSymbols(Color color, Brush brush, int size) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/paint/toolBox/shape/Shape.java]: Bad method name. This method just need to print symbol by given length.
Method MUST have ONLY 2 parameters - symbol, and the length

File: src/main/java/am/hsp/apps/paint/Paint.java [hunk 14]
```diff
@@ -0,0 +1,80 @@
+package main.java.am.hsp.apps.paint;
+
+import main.java.am.hsp.apps.paint.toolBox.ToolBox;
+import main.java.am.hsp.apps.paint.toolBox.brush.Brush;
+import main.java.am.hsp.apps.paint.toolBox.color.Color;
+import main.java.am.hsp.apps.paint.toolBox.shape.Shape;
+
+/**
+ * Class Paint
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/paint/Paint.java]: YOu didn't fix my comment related your bad comments,,,

File: src/main/java/am/hsp/apps/paint/PaintArea.java [hunk 15]
```diff
@@ -0,0 +1,24 @@
+package main.java.am.hsp.apps.paint;
+
+import main.java.am.hsp.apps.paint.toolBox.brush.Brush;
+import main.java.am.hsp.apps.paint.toolBox.color.Color;
+import main.java.am.hsp.apps.paint.toolBox.shape.Shape;
+
+/**
+ * Class PaintArea
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/paint/PaintArea.java]: same

File: src/main/java/am/hsp/apps/paint/exception/InvalidBrushException.java [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+    /**
+     * 
+     */
+    private static final long serialVersionUID = 1L;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/paint/exception/InvalidBrushException.java]: please generate **auto generated** serialVersionId - not just 1L (remove this serial version id, click on the warning, and chose an option "auto generated serial version id ")

File: src/main/java/am/hsp/apps/paint/toolBox/shape/Diamond.java [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+ */
+public class Diamond extends Shape {
+
+    public static final int DIAMOND_ID = 3;
+
+    /**
+     * Constructor for initializing
+     */
+    public Diamond() {
+        super.id = DIAMOND_ID;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/paint/toolBox/shape/Diamond.java]: Why super??? You need to have id as protected variable, and just initialize here as an instance variable of Diamond class - not of Shape class...

**Comment by hsp-arman** [comment]
Dear Harut ,
sorry for  this bugs.  Previously mistakes were directed, but for some unknown reason they did not survive ․ All already directed, including some more ..

Regards
 Arman

**Pull request merged**